csi/src/findmnt.rs

```diff
@@ -0,0 +1,223 @@
+            }
+        } else {
+            //FIXME: key:value pairs are discarded if the value is not a
+            // string.
+            error!("value for {} is not a string", key);
+        }
+    }
+    hmap
+}
+
+/// This function recurses over the de-serialised JSON returned by findmnt,
+/// finding entries which have key-pair's matching the filter key-pair,
+/// and populates a vector with the values for the item_key.
+///
+/// For Mayastor usage the assumptions made on the structure are:
+///  1. An object has keys named "target" and "source" for a mount point.
+///  2. An object may contain nested arrays of objects.
+///
+/// The search is deliberately generic (and hence slower) in an attempt to
+/// be more robust to future changes in findmnt.
+fn filter_findmnt(
+    json_val: &serde_json::value::Value,
+    filter: &FindmntFilter,
+    results: &mut Vec<HashMap<String, String>>,
+) {
+    if let Some(json_array) = json_val.as_array() {
+        for jsonvalue in json_array {
+            filter_findmnt(&jsonvalue, filter, results);
+        }
+    }
+    if let Some(json_map) = json_val.as_object() {
+        if let Some(value) = json_map.get(filter.key) {
+            if filter == value {
+                results.push(jsonmap_to_hashmap(json_map));
+            }
+        }
+        // If the object has arrays, then the assumption is that they are arrays
+        // of objects.
+        for (_, jsonvalue) in json_map {
+            if jsonvalue.is_array() {
+                filter_findmnt(jsonvalue, filter, results);
+            }
+        }
+    }
+}
+
+/// findmnt executable name.
+const FINDMNT: &str = "findmnt";
+/// findmnt arguments, we only want source, target and filesystem type fields.
+const FINDMNT_ARGS: [&str; 3] = ["-J", "-o", "SOURCE,TARGET,FSTYPE"];
+
+/// Execute the Linux utility findmnt, collect the json output,
+/// invoke the filter function and return the filtered results.
+fn findmnt(
+    params: FindmntFilter,
+) -> Result<Vec<HashMap<String, String>>, DeviceError> {
+    let output = Command::new(FINDMNT).args(&FINDMNT_ARGS).output()?;
+    if output.status.success() {
+        let json_str = String::from_utf8(output.stdout)?;
+        let json: Value = serde_json::from_str(&json_str)?;
+        let mut results: Vec<HashMap<String, String>> = Vec::new();
+        filter_findmnt(&json, &params, &mut results);
+        Ok(results)
+    } else {
+        Err(DeviceError {
+            message: String::from_utf8(output.stderr)?,
+        })
+    }
+}
+
+/// Use the Linux utility findmnt to find the name of the device mounted at a
+/// directory or block special file, if any.
+/// mount_path is the path a device is mounted on.
+pub(crate) fn findmnt_get_devicepath(
+    mount_path: &str,
+) -> Result<Option<String>, DeviceError> {
+    let tgt_filter = FindmntFilter {
+        key: TARGET_KEY,
+        value: mount_path,
+    };
+    match findmnt(tgt_filter) {
```

Comment:
if you use ? here you can dispense of the Err(e) => Err(e) thing

Reply:
Done.